model/map/src/main/java/org/keycloak/models/map/client/MapClientProvider.java

```diff
@@ -0,0 +1,316 @@
+                return clientRegisteredNodesStore.computeIfAbsent(entity.getId(), k -> new ConcurrentHashMap<>());
+            }
+
+            @Override
+            public void registerNode(String nodeHost, int registrationTime) {
+                Map<String, Integer> value = getRegisteredNodes();
+                value.put(nodeHost, registrationTime);
+            }
+
+            @Override
+            public void unregisterNode(String nodeHost) {
+                getRegisteredNodes().remove(nodeHost);
+            }
+
+        };
+    }
+
+    private Predicate<MapClientEntity> entityRealmFilter(RealmModel realm) {
+        if (realm == null || realm.getId() == null) {
+            return MapClientProvider.ALWAYS_FALSE;
+        }
+        String realmId = realm.getId();
+        return entity -> Objects.equals(realmId, entity.getRealmId());
+    }
+
+    @Override
+    public List<ClientModel> getClients(RealmModel realm, Integer firstResult, Integer maxResults) {
+        Stream<ClientModel> s = getClientsStream(realm);
+        if (firstResult >= 0) {
+            s = s.skip(firstResult);
+        }
+        if (maxResults >= 0) {
+            s = s.limit(maxResults);
+        }
+        return s.collect(Collectors.toList());
+    }
+
+    private Stream<MapClientEntity> getNotRemovedUpdatedClientsStream() {
+        Stream<MapClientEntity> updatedAndNotRemovedClientsStream = clientStore.entrySet().stream()
+          .map(tx::getUpdated)    // If the client has been removed, tx.get will return null, otherwise it will return me.getValue()
+          .filter(Objects::nonNull);
+        return Stream.concat(tx.createdValuesStream(clientStore.keySet()), updatedAndNotRemovedClientsStream);
+    }
+
+//    @Override
+    public Stream<ClientModel> getClientsStream(RealmModel realm) {
+        return getNotRemovedUpdatedClientsStream()
+          .filter(entityRealmFilter(realm))
+          .sorted(COMPARE_BY_CLIENT_ID)
+          .map(entityToAdapterFunc(realm))
+        ;
+    }
+
+    @Override
+    public List<ClientModel> getClients(RealmModel realm) {
+        return getClientsStream(realm).collect(Collectors.toList());
+    }
+
+    @Override
+    public ClientModel addClient(RealmModel realm, String id, String clientId) {
+        final UUID entityId = id == null ? UUID.randomUUID() : UUID.fromString(id);
+
+        if (clientId == null) {
+            clientId = entityId.toString();
+        }
+
+        LOG.tracef("addClient(%s, %s, %s)%s", realm, id, clientId, getShortStackTrace());
+
+        MapClientEntity entity = new MapClientEntity(entityId, realm.getId());
+        entity.setClientId(clientId);
+        entity.setEnabled(true);
+        entity.setStandardFlowEnabled(true);
+        if (tx.get(entity.getId(), clientStore::get) != null) {
+            throw new ModelDuplicateException("Client exists: " + id);
+        }
+        tx.putIfAbsent(entity.getId(), entity);
```

Comment:
Why `putIfAbsent()`? We checked there is no client with the same id.

Reply:
Because this needs to be done in a transaction.

Reply:
MapKeycloakTransaction supports also `put()`

Reply:
Yup, but the transaction is only committed after the request is almost processed. The `get` above provides a preliminary check during processing the request - a fail-fast functionality. If a concurrent transaction (e.g. concurrent request) adds the same client, this transaction should not force replacing the client.

Reply:
Ok I see, thanks